audio_service/lib/audio_service.dart

```diff
@@ -914,40 +916,76 @@ class AudioService {
     return handler;
   }
 
+  static CancelableOperation<Object?>? _artFetchOperation;
+  static late double _debugCurrentFetchId;
   static Future<void> _observeMediaItem() async {
     await for (var mediaItem in _handler.mediaItem) {
       if (mediaItem == null) continue;
+      _artFetchOperation?.cancel();
       final artUri = mediaItem.artUri;
       if (artUri != null) {
-        // We potentially need to fetch the art.
-        String? filePath;
+        /// Sends media item to the platform.
+        /// We potentially need to fetch the art before that.
+        Future<void> _sendToPlatform(String? filePath) async {
+          final extras =
+              Map<String, dynamic>.of(mediaItem.extras ?? <String, dynamic>{});
+          extras['artCacheFile'] = filePath;
+          final platformMediaItem = mediaItem.copyWith(extras: extras);
+          // Show the media item after the art is loaded.
+          await _platform.setMediaItem(
+              SetMediaItemRequest(mediaItem: platformMediaItem._toMessage()));
+        }
+
         if (artUri.scheme == 'file') {
-          filePath = artUri.toFilePath();
+          await _sendToPlatform(artUri.toFilePath());
         } else {
-          final fileInfo =
-              await cacheManager.getFileFromMemory(artUri.toString());
-          filePath = fileInfo?.file.path;
-          if (filePath == null) {
-            // We haven't fetched the art yet, so show the metadata now, and again
-            // after we load the art.
-            await _platform.setMediaItem(
-                SetMediaItemRequest(mediaItem: mediaItem._toMessage()));
-            // Load the art
-            filePath = await _loadArtwork(mediaItem);
-            // If we failed to download the art, abort.
-            if (filePath == null) continue;
-            // If we've already set a new media item, cancel this request.
-            // XXX: Test this
-            //if (mediaItem != _handler.mediaItem.value) continue;
-          }
+          // On each step of operation it should be ensured there is only
+          // one operation running and previous are cancelled.
+          late final double id;
+          assert(() {
+            id = math.Random().nextDouble();
+            _debugCurrentFetchId = id;
+            return true;
+          }());
+
+          _artFetchOperation =
+              CancelableOperation<String?>.fromFuture(() async {
+            // Try to load a cached file from memory.
+            final fileInfo =
+                await cacheManager.getFileFromMemory(artUri.toString());
+            return fileInfo?.file.path;
+          }())
+                  .then(
+            (filePath) {
+              assert(id == _debugCurrentFetchId);
+              if (filePath != null) {
+                // If we successfully downloaded the art call to platform.
+                _sendToPlatform(filePath);
+              } else {
+                _artFetchOperation = CancelableOperation.fromFuture(
```

Comment:
Correct me if I'm wrong, but I don't think you need to manually update `_artFetchOperation` for each link in the chain. Or at least that seems to be the case when I look at the dart project test cases.

Reply:
There's a branching here - if we up in `true` condition, we don't want to continue the operation.

Instead of creating new cancelable operation in `else`, you could write this entire thing in a single chain of `then`s, but then you would need to add a very easy to miss call to `cancel` in the true condition above. In my opinion this would be harder to read, but it is more a preference of the style.

Reply:
I'm still trying to get my head around the way `CancelableOperation` works, so actually at the moment it's hard to read both ways. What I think would be ideal is if `then()` chained them together in such a way that you don't have to keep track of the chain links yourself. If `CancelableOperation` doesn't already do this, then I would wonder why not (and would probably create my own abstraction around it to do it), but those test cases seemed to indicate the desired behaviour would work. I'll probably have to do some tests myself..

Reply:
`CancelableOperation` confuses me completely. This program aborts early in `op.value` without error:

```dart
Future<void> main(List<String> args) async {
  try {
    final op = CancelableOperation.fromFuture((() async {
      print('one');
      await Future.delayed(Duration(seconds: 2));
      return 1;
    })())
        .then((value) async {
      print('two');
      await Future.delayed(Duration(seconds: 2));
      return 2;
    }).then((value) async {
      print('three');
      await Future.delayed(Duration(seconds: 2));
      return 3;
    });
    await Future.delayed(Duration(seconds: 1));
    op.cancel();
    print('cancelled');
    print('value: ${await op.value}');
    print('end of try');
  } catch (e) {
    print('Error: $e');
  }
  print('foo');
  await Future.delayed(Duration(seconds: 10));
  print('bar');
}
```

Reply:
You are awaiting the value, so at the moment of you getting it, you are actually done waiting for the first future for 2 seconds, yes, it doesn't cancel it. But all the queued `then`s are not executed and that's the idea behind the API. Is there anything in my code that concerns you?

Reply:
Or if I missed something you wanted to say with this example please explain it, I didn't actually run it

Reply:
Add the moment I'm just trying to fully understand how this class works in order to the understand code that uses it. My example aborts early. It doesn't do that long delay at the end after the try catch, it aborts before then. Isn't that strange?

Reply:
Can you show me what is the expected vs actual results are? I'm not sure I understand

Reply:
> it aborts before then. Isn't that strange?

I'm confused, because I would expect exactly that

Reply:
The only normal means I know to abort is an exception or an `exit` call. It's not an exception because I caught the exception. If that were the case, it should have flowed through to the last 3 lines starting at `foo`. The behaviour I'm seeing is consistent with `exit` but that wouldn't be expected behaviour of this API. Other than that, the abnormal explanation could be a bug in the Dart runtime.

By what means do you think this example exits the `main` method without reaching the lines after the try catch?

Reply:
I should add that `return` is another normal case to abort, but the `main` method above doesn't return, only the anonymous functions return.

Reply:
Ahh, as I said, I didn't run the sample, now I see. I was confused by this at first too. Please see https://github.com/dart-lang/async/issues/163

If you want it to return there, you need to use valueOrCancellation

Reply:
To be clear, I totally forgot what `value` actually does, and due to lack of possibility to run it, assumed the wrong thing above